File: utils/metadata_extractor.py
import logging
from typing import Dict, Optional
from mutagen.flac import FLAC
from mutagen import MutagenError 
from file_handler import FileHandler

logger = logging.getLogger(__name__)


class MetadataExtractor:
    """Extract metadata from FLAC files"""
    
    def extract_from_flac(self, file_path: str , file_handler: FileHandler) -> Dict[str, Optional[str]]:
        """
        Extract metadata from FLAC file
        
        Args:
            file_path: Path to the FLAC file
            
        Returns:
            Dictionary containing extracted metadata
        """
        try:
            audio = FLAC(file_path)
            if audio is None:
                logger.error(f"Could not read FLAC file: {file_path}")
                return {}
             
            cover_path = self.extract_flac_cover_art(audio , file_handler)

            metadata = {
                "duration_seconds": int(audio.info.length),
                "song_title": self._get_tag(audio, "TITLE"),
                "artist_main": self._get_tag(audio, "ARTIST"),
                "album_title": self._get_tag(audio, "ALBUM"),
                "composer": self._get_tag(audio, "COMPOSER"),
                "band_name": self._get_tag(audio, "ALBUMARTIST"),
                "label": self._get_tag(audio, "LABEL"),
                "year_recorded": self._get_year(audio, "DATE"),
                "year_released": self._get_year(audio, "ORIGINALDATE"),
                "song_order": self._get_track_number(audio),
                "instrument": self._get_tag(audio, "INSTRUMENT"),
                "comments": self._get_tag(audio, "COMMENT"),
                "other_artist_playing": self._get_tag(audio, "PERFORMER"),
                "musicbrainz_track_id": self._get_tag(audio, "MUSICBRAINZ_TRACKID"),
                "musicbrainz_album_id": self._get_tag(audio, "MUSICBRAINZ_ALBUMID"),
                "musicbrainz_artist_id": self._get_tag(audio, "MUSICBRAINZ_ARTISTID"),
            }
            logger.info(metadata)
            
            # Clean up None values
            return {k: v for k, v in metadata.items() if v is not None}
            
        except MutagenError as e:
            logger.error(f"Error reading FLAC metadata: {e}")
            return {}
        except Exception as e:
            logger.error(f"Unexpected error extracting metadata: {e}")
            return {}
    
    def _get_tag(self, audio: FLAC, tag: str) -> Optional[str]:
        """Get a single tag value"""
        values = audio.get(tag)
        if values and len(values) > 0:
            return str(values[0]).strip()
        return None
    
    def extract_flac_cover_art(self, audio: FLAC, file_handler: FileHandler):
        """
        Extracts the front cover art from a FLAC file and saves it as an image.

        Args:
            file_path (str): The path to the FLAC file.
            output_image_path (str): The path where the extracted image will be saved.
        """
        try:  
            # Iterate through embedded pictures
            for pic in audio.pictures: 
                file_name = 'cover_' + self._get_tag(audio, "ARTIST")+'_'+self._get_tag(audio, "ALBUM")
                file_name= file_name.replace(' ' , '_') +'.jpg'
                relative_path = file_handler.save_album_cover_from_flac(pic.data, file_name)
                return relative_path
                     
            print("No front cover art found in the FLAC file.")

        except Exception as e:
            print(f"Error extracting cover art: {e}")

    def _get_year(self, audio: FLAC, tag: str) -> Optional[int]:
        """Extract year from date tag"""
        date_str = self._get_tag(audio, tag)
        if date_str:
            try:
                # Handle various date formats (YYYY, YYYY-MM-DD, etc.)
                year = int(date_str.split("-")[0])
                return year if 1900 <= year <= 2100 else None
            except (ValueError, IndexError):
                logger.warning(f"Could not parse year from: {date_str}")
        return None
    
    def _get_track_number(self, audio: FLAC) -> Optional[int]:
        """Extract track number"""
        track_str = self._get_tag(audio, "TRACKNUMBER")
        if track_str:
            try:
                # Handle "1/12" format
                return int(track_str.split("/")[0])
            except (ValueError, IndexError):
                logger.warning(f"Could not parse track number from: {track_str}")
        return None

File: utils/metadataExtractor.js
const fs = require('fs');
const path = require('path');
const { parseBuffer } = require('music-metadata');

class MetadataExtractor {
    /**
     * Extract metadata from FLAC file
     * @param {string} filePath - Path to the FLAC file
     * @param {Object} fileHandler - File handler instance
     * @returns {Object} Dictionary containing extracted metadata
     */
    async extractFromFlac(filePath, fileHandler) {
        try {
            const buffer = fs.readFileSync(filePath);
            const metadata = await parseBuffer(buffer, 'audio/flac');
            
            if (!metadata) {
                console.error(`Could not read FLAC file: ${filePath}`);
                return {};
            }

            // Extract cover art
            const coverPath = await this.extractFlacCoverArt(metadata, fileHandler);

            const extractedMetadata = {
                duration_seconds: Math.floor(metadata.format.duration || 0),
                song_title: this._getTag(metadata, 'title'),
                artist_main: this._getTag(metadata, 'artist'),
                album_title: this._getTag(metadata, 'album'),
                composer: this._getTag(metadata, 'composer'),
                band_name: this._getTag(metadata, 'albumartist'),
                label: this._getTag(metadata, 'label'),
                year_recorded: this._getYear(metadata, 'date'),
                year_released: this._getYear(metadata, 'originaldate'),
                song_order: this._getTrackNumber(metadata),
                instrument: this._getTag(metadata, 'instrument'),
                comments: this._getTag(metadata, 'comment'),
                other_artist_playing: this._getTag(metadata, 'performer'),
                musicbrainz_track_id: this._getTag(metadata, 'musicbrainz_trackid'),
                musicbrainz_album_id: this._getTag(metadata, 'musicbrainz_albumid'),
                musicbrainz_artist_id: this._getTag(metadata, 'musicbrainz_artistid'),
            };

            console.log('Extracted metadata:', extractedMetadata);

            // Clean up null/undefined values
            return Object.fromEntries(
                Object.entries(extractedMetadata).filter(([_, v]) => v != null)
            );

        } catch (error) {
            console.error(`Error reading FLAC metadata: ${error.message}`);
            return {};
        }
    }

    /**
     * Get a single tag value
     * @param {Object} metadata - Parsed metadata object
     * @param {string} tag - Tag name
     * @returns {string|null} Tag value or null
     */
    _getTag(metadata, tag) {
        const common = metadata.common || {};
        const value = common[tag.toLowerCase()];
        
        if (Array.isArray(value) && value.length > 0) {
            return value[0].toString().trim();
        } else if (value) {
            return value.toString().trim();
        }
        return null;
    }

    /**
     * Extract cover art from FLAC file and save it
     * @param {Object} metadata - Parsed metadata object
     * @param {Object} fileHandler - File handler instance
     * @returns {string|null} Relative path to saved cover or null
     */
    async extractFlacCoverArt(metadata, fileHandler) {
        try {
            const pictures = metadata.common?.picture;
            if (!pictures || pictures.length === 0) {
                console.log('No cover art found in the FLAC file.');
                return null;
            }

            // Use the first picture (usually front cover)
            const picture = pictures[0];
            const artist = this._getTag(metadata, 'artist') || 'unknown';
            const album = this._getTag(metadata, 'album') || 'unknown';
            
            const fileName = `cover_${artist}_${album}`.replace(/\s+/g, '_') + '.jpg';
            const relativePath = await fileHandler.saveAlbumCoverFromFlac(picture.data, fileName);
            
            return relativePath;

        } catch (error) {
            console.error(`Error extracting cover art: ${error.message}`);
            return null;
        }
    }

    /**
     * Extract year from date tag
     * @param {Object} metadata - Parsed metadata object
     * @param {string} tag - Tag name
     * @returns {number|null} Year or null
     */
    _getYear(metadata, tag) {
        const dateStr = this._getTag(metadata, tag);
        if (dateStr) {
            try {
                const year = parseInt(dateStr.split('-')[0]);
                return (year >= 1900 && year <= 2100) ? year : null;
            } catch (error) {
                console.warn(`Could not parse year from: ${dateStr}`);
            }
        }
        return null;
    }

    /**
     * Extract track number
     * @param {Object} metadata - Parsed metadata object
     * @returns {number|null} Track number or null
     */
    _getTrackNumber(metadata) {
        const trackStr = this._getTag(metadata, 'track');
        if (trackStr) {
            try {
                // Handle "1/12" format
                return parseInt(trackStr.split('/')[0]);
            } catch (error) {
                console.warn(`Could not parse track number from: ${trackStr}`);
            }
        }
        return null;
    }
}

module.exports = MetadataExtractor;